### PR TITLE
Add validations to PersonalDetailsForm

### PR DESCRIPTION
**Pull request by tvararu**
### Context

We need to make sure that users don't submit invalid data when they add their personal details to an application form.

### Changes proposed in this pull request

- Add validations, with unit tests
- Add an end to end test for a basic case (will be split out once `PersonalDetailsForm` is finished)

### Guidance to review

I didn't opt to make `date_of_birth` its own separate `Validator` because I don't expect it to be reused, and also naming it and figuring out options to pass in (`{ future: true }` etc) were giving me "overengineering" vibes so I decided to just leave it as a simple class method.

The constant `raise`ing in the date of birth stuff is a bit `GOTO`y and maybe could be improved?

I've split out some PRs for some of the commits on this in order to make this smaller:

- [x] `WordCountValidator` https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/302
- [x] `NATIONALITIES` https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/305
- [x] Replacing `nationalities` with `first_nationality` and `second_nationality` https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/306

### After screenshot

Screenshot 2019-10-11 at 12 22 47

### Link to Trello card

[122 - Allow users to fill in "Personal details" with validation](https://trello.com/c/LTQDGPQh/122-allow-users-to-fill-in-personal-details-with-validation)

### CODE REVIEW

File: app/models/candidate_interface/personal_details_form.rb [hunk 1]
```diff
@@ -8,8 +8,37 @@ class PersonalDetailsForm
...
+    validates :english_language_details, :other_language_details,
+              word_count: { maximum: 200 }
+
     def name
       "#{first_name} #{last_name}"
     end
+
+    def date_of_birth
+      Date.new(*[year, month, day].map(&:to_i))
+    rescue ArgumentError, NoMethodError
```
Comment (tijmenb) [inline app/models/candidate_interface/personal_details_form.rb]: Missed this in the review - I'd rather not `rescue` these errors, is there an alternative?

File: app/models/candidate_interface/personal_details_form.rb [hunk 2]
```diff
@@ -8,8 +8,37 @@ class PersonalDetailsForm
...
+      nil
+    end
+
+    def date_of_birth_cannot_be_in_the_future
+      raise 'Invalid date' if [year, month, day].any?(&:blank?)
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/models/candidate_interface/personal_details_form.rb]: This line is not testing whether the DOB is in the future — does it belong in its own method? 

(I agree with @tijmenb fwiw)

File: app/models/candidate_interface/personal_details_form.rb [hunk 3]
```diff
@@ -8,8 +8,37 @@ class PersonalDetailsForm
...
+
+      errors.add(:date_of_birth, I18n.t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.date_of_birth.future')) if date_of_birth > Date.today
+    rescue RuntimeError, NoMethodError
+      errors.add(:date_of_birth, I18n.t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.date_of_birth.invalid'))
```
Comment (duncanjbrown) [inline app/models/candidate_interface/personal_details_form.rb]: I feel sure there's a way to do this without writing out the fully-qualified path of the translation. 🐇

Comment (duncanjbrown) [reply app/models/candidate_interface/personal_details_form.rb]: Maybe pass `:invalid` in the second place? https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-add

File: spec/models/candidate_interface/personal_details_form_spec.rb [hunk 4]
```diff
@@ -8,4 +8,52 @@
...
+    okay_text = Faker::Lorem.sentence(word_count: 200)
+    long_text = Faker::Lorem.sentence(word_count: 201)
+
+    it { is_expected.to allow_value(okay_text).for(:english_language_details) }
+    it { is_expected.not_to allow_value(long_text).for(:english_language_details) }
+
+    it { is_expected.to allow_value(okay_text).for(:other_language_details) }
+    it { is_expected.not_to allow_value(long_text).for(:other_language_details) }
+
+    context 'date of birth' do
```
Comment (duncanjbrown) [inline spec/models/candidate_interface/personal_details_form_spec.rb]: This feels more like a `describe` to me?

File: spec/models/candidate_interface/personal_details_form_spec.rb [hunk 5]
```diff
@@ -8,4 +8,52 @@
...
+      it 'validates that the date is valid' do
```
Comment (duncanjbrown) [inline spec/models/candidate_interface/personal_details_form_spec.rb]: ` it "is invalid if not well-formed"`?

File: app/models/candidate_interface/personal_details_form.rb [hunk 6]
```diff
@@ -8,8 +8,37 @@ class PersonalDetailsForm
...
+      Date.new(*[year, month, day].map(&:to_i))
+    rescue ArgumentError, NoMethodError
+      nil
+    end
+
+    def date_of_birth_cannot_be_in_the_future
+      raise 'Invalid date' if [year, month, day].any?(&:blank?)
+
+      errors.add(:date_of_birth, I18n.t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.date_of_birth.future')) if date_of_birth > Date.today
+    rescue RuntimeError, NoMethodError
```
[the remark this answers is not in the document]
Comment (tvararu) [reply app/models/candidate_interface/personal_details_form.rb]: Agree 👍

File: app/models/candidate_interface/personal_details_form.rb (hunk 3 above)
Comment (tvararu) [reply app/models/candidate_interface/personal_details_form.rb]: Nice, we'll give that a go!

File: app/models/candidate_interface/personal_details_form.rb (hunk 1 above)
Comment (tvararu) [reply app/models/candidate_interface/personal_details_form.rb]: Fix'd using `date_valid?`, also added more tests!

**Comment by tvararu** [comment]
@tijmenb @duncanjbrown have refactored the `date_of_birth` stuff and it's a lot cleaner, give it a 👀 !

File: app/models/candidate_interface/personal_details_form.rb [hunk 7]
```diff
@@ -8,8 +8,39 @@ class PersonalDetailsForm
...
+
+    def date_of_birth
+      date_args = [year, month, day].map(&:to_i)
+      if Date.valid_date?(*date_args)
+        Date.new(*date_args)
+      end
+    end
+
+    def date_of_birth_valid
+      errors.add(:date_of_birth, :invalid) if date_of_birth.nil?
```
Comment (duncanjbrown) [inline app/models/candidate_interface/personal_details_form.rb]: 😗👌

Comment (duncanjbrown) [reply app/models/candidate_interface/personal_details_form.rb]: (tbc, that was supposed to be a person kissing their fingers, not a kiss followed by an ok_hand, which would be weird)

File: spec/system/candidate_interface/candidate_entering_personal_details_spec.rb [hunk 8]
```diff
@@ -77,4 +87,12 @@ def then_i_can_check_my_answers
   def then_i_can_see_my_application
     expect(page).to have_content t('page_titles.application_form')
   end
+
+  def then_i_should_see_validation_errors
+    expect(page).to have_content t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.first_name.blank')
+    expect(page).to have_content t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.last_name.blank')
+    expect(page).to have_content t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.date_of_birth.invalid')
+    expect(page).to have_content t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.first_nationality.blank')
+    expect(page).to have_content t('activemodel.errors.models.candidate_interface/personal_details_form.attributes.english_main_language.blank')
```
Comment (duncanjbrown) [inline spec/system/candidate_interface/candidate_entering_personal_details_spec.rb]: Could this be made more readable using `scope:`? Eg

```
scope = 'activemodel.errors.models.candidate_interface/personal_details_form.attributes'
expect(page).to have_content t('first_name.blank', scope: scope)
expect(page).to have_content t('last_name.blank', scope: scope)
```

Comment (tvararu) [reply spec/system/candidate_interface/candidate_entering_personal_details_spec.rb]: Oh yes <3

**Pull request merged**